docs: update Emoji ID (#48)
tari-project/rfcs: src/RFC-0152_EmojiId.md

## Goals

> This document describes the specification for Emoji Ids. Emoji Ids are encoded node ids used for humans to verify peer node addresses easily.

Reviewer: ```suggestion
This document describes the specification for Emoji Ids. Emoji Ids are encoded node ids used for humans to verify peer node addresses easily and for machines to verify that the address is being used in the correct context.
```

Tari [Communication Node]s are identified on the network via their [Node ID]; which in turn are derived from the node's
public key. Both the node id and public key are simple large integer numbers.

The most common practice for human beings to copy large numbers in cryptocurrency software is scanning a QR code or copying and pasting a value from one application to another. These numbers are typically encoded using hexadecimal or Base58
encoding. The user will then typically scan (parts) of the string by eye to ensure that the value was transferred
correctly.

> For Tari, we propose encoding values, the node ID and network identifier in particular, for Tari, using emojis. The advantages of this approach are:

Reviewer: The network id is masking the checksum, so is not encoded.

* ...
* 🦊 => 255

> The emoji SHOULD be selected such that.

Reviewer 2: Probably want a colon `:` here instead.

* Similar-looking emoji are excluded from the map. e.g. Neither 😁 or 😄 should be included. Similarly, the Irish and
  Côte d'Ivoire flags look very similar, and both should be excluded.
* Modified emoji (skin tones, gender modifiers) are excluded. Only the "base" emoji is considered.

### Encoding

The selection of an alphabet with 256 symbols means there is a direct mapping between bytes and emoji. For each byte
in the input data to be encoded, map the byte to the corresponding emoji using the emoji map. The resulting
concatenation of emoji characters is the emoji string.

### Emoji ID definition

The emoji ID is a string of 33 symbols from the emoji alphabet. It uses this bitmap:

```text
+-----------------------------+--------------------------+

> |  Node public key (256 bits) | EncodedChecksum (8 bits) |

Reviewer 2: I wouldn't use the term "encoded" for this, as I don't think it really conveys the nature of the modification. Maybe something like "masked" instead?

+-----------------------------+--------------------------+
```

>  The emoji ID is calculated from a node public key serialized as 32 bytes (`B`) as follows:

Reviewer 2: Should mention that the 8-bit network encoding is part of the algorithm input here, along with the public key. Otherwise it's not clear where it comes into play.

* Use the [DammSum](https://github.com/cypherstack/dammsum) algorithm with `k = 8` and `m = 32` to compute an 8-bit

> encoded checksum `eC` from `B`.

Reviewer 2: I'd be more specific about this, to avoid any confusion:
"... to compute an 8-bit masked checksum `C'` from `B` (described below)"

Otherwise, the casual reader might assume this is just the standard DammSum checksum.

Reviewer 2: I would also avoid the use of multiple symbols for the variable name `eC`, but that's just a nit and a personal preference.

* Encode `B` into an emoji string.
* Encode `eC` into an emoji string.
* Concatenate `B` and `eC` as the emoji ID.

### Checksum 

> For the checksum we use the [DammSum](https://github.com/cypherstack/dammsum) algorithm with `k = 8` and `m = 32` to compute an 8-bit

Reviewer 2: "For the masked checksum we use..."

encoded checksum `C` from the public key `B`.

> To verify that the network that the public key comes from is the same one, we encode the checksum with the network. 

Reviewer 2: I would use a term like "assert" instead of "verify", since verification seems to imply part of the decoding and validation process.

We calculate the encoded checksum `eC` as `C XOR N` where `N` is the network defined as 8-bits.

### Decoding

> One can extract the node public key from an emoji ID as follows:

Reviewer 2: "... from an emoji ID and expected network `N` as follows:"

* ...
* 🦊 => 255

> The emoji SHOULD be selected such that:

Reviewer: Change language to past continuous tense.

```suggestion
The emoji have been selected such that:
```

* Similar-looking emoji are excluded from the map. e.g. Neither 😁 or 😄 should be included. Similarly, the Irish and

>   Côte d'Ivoire flags look very similar, and both should be excluded.

Reviewer: May as add the table of emoji here and their encoding values.

### Encoding

The selection of an alphabet with 256 symbols means there is a direct mapping between bytes and emoji. For each byte
in the input data to be encoded, map the byte to the corresponding emoji using the emoji map. The resulting
concatenation of emoji characters is the emoji string.

### Emoji ID definition

The emoji ID is a string of 33 symbols from the emoji alphabet. It uses this bitmap:

```text
+-----------------------------+--------------------------+
|  Node public key (256 bits) | MaskedChecksum (8 bits)  |
+-----------------------------+--------------------------+
```

 The emoji ID is calculated from a node public key serialized as 32 bytes (`B`) as follows:

* Use the [DammSum](https://github.com/cypherstack/dammsum) algorithm with `k = 8` and `m = 32` to compute an 8-bit
masked checksum `'C` from `B` and network `N`.
* Encode `B` into an emoji string.
* Encode `'C` into an emoji string.
* Concatenate `B` and `'C` as the emoji ID.

#### Checksum effectiveness
It is important to note that by masking the checksum we technically reduce the effectiveness of the checksum. These emoji ids wont

> however be sent over the wire in raw form abd its far more likely that they wil just be copied and thus its important to verify the

Reviewer 3: ```suggestion
however, be sent over the wire in raw form and its far more likely that they will just be copied thus it's important to verify the
```

> network of the wallet.

Reviewer: Specifically, if you want to get technical, there's a ~4/256 = 1/64 chance that a transcription mistake leads to the incorrect network/version combination matching the checksum (assuming 4 ver/net combos).
